src/vcpu.rs

```diff
@@ -0,0 +1,204 @@
+// Copyright 2018-2019 CrowdStrike, Inc.
+// SPDX-License-Identifier: Apache-2.0 or MIT
+//
+// Portions Copyright 2018 Cloudbase Solutions Srl
+// SPDX-License-Identifier: Apache-2.0 or MIT
+//
+// Portions Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+//
+// Portions Copyright 2017 The Chromium OS Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the THIRD-PARTY file.
+
+//! Trait to access the functionality of a virtual CPU (vCPU).
+
+use std::{io, result};
+
+#[cfg(any(target_arch = "arm", target_arch = "aarch64"))]
+pub use arm::VcpuInit;
+
+#[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+pub use x86_64::{
+    StandardRegisters, SpecialRegisters, FpuState, MsrEntries, MsrEntry, 
+    CpuId, LapicState
+};
+
+///
+/// Reasons for vCPU exits.
+///
+/// Exit reasons are drawn a superset of exit reasons for various VMMs.
+///
+#[derive(Debug)]
+pub enum VcpuExit<'a> {
+    None,
+    MemoryAccess,
+    IoPortAccess,
+    UnrecoverableException,
+    InvalidVpRegisterValue,
+    UnsupportedFeature,
+    MsrAccess,
+    Cpuid,
+    Canceled,
+
+    /// An out port instruction was run on the given port with the given data.
+    IoOut(u16 /* port */, &'a [u8] /* data */),
+    /// An in port instruction was run on the given port.
+    ///
+    /// The given slice should be filled in before Vcpu::run() is called again.
+    IoIn(u16 /* port */, &'a mut [u8] /* data */),
+    /// A read instruction was run against the given MMIO address.
+    ///
+    /// The given slice should be filled in before Vcpu::run() is called again.
+    /// is called again.
+    MmioRead(u64 /* address */, &'a mut [u8]),
+    /// A write instruction was run against the given MMIO address with the given data.
+    MmioWrite(u64 /* address */, &'a [u8]),
+    Unknown,
+    Hypercall,
+    Debug,
+    Shutdown,
+    FailEntry,
+    Intr,
+    SetTpr,
+    TprAccess,
+    S390Sieic,
+    S390Reset,
+    Dcr,
+    Nmi,
+    InternalError,
+    Osi,
+    PaprHcall,
+    S390Ucontrol,
+    Watchdog,
+    S390Tsch,
+    Epr,
+    SystemEvent,
+    S390Stsi,
+    Hyperv,
+
+    Hlt,
+    IoapicEoi,
+    Exception,
+    IrqWindowOpen,
+}
+
+pub type Result<T> = result::Result<T, io::Error>;
```

Comment:
Please add the documentation from kvm-ioctls:
```
/// A specialized `Result` type for KVM ioctls.
///
/// This typedef is generally used to avoid writing out io::Error directly and
/// is otherwise a direct mapping to Result.
```

It needs to be adapted so that it doesn't mention KVM.

Reply:
Done